readme.md
# Get System Report Documentation
(c) 2016-2017 Michal Zobec, ZOBEC Consulting. All Rights Reserved.  
web: www.michalzobec.cz, mail: michal@zobec.net  
License: Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)  
https://creativecommons.org/licenses/by-sa/4.0/

# HISTORY
Release notes can find in file [ChangeLog.md](https://github.com/michalzobec/PowerShell/blob/master/Get-SystemReport/changelog.md).

# FEATURES
- integrated basic system report: Hardware Information, Windows Information, Network Information;
- Windows product key (in Windows Information Section) is now possibly not include to report;
- Windows Information: uptime, last boot time;
- Windows Firewall Status;
- AntiVirus Status;
- Windows Task Scheduler Failed Tasks;
- collect of windows eventlog from defined time (from 24 hours back to actual time);
- collect of windows eventlog from selected logs: Application, System;
- collect of windows eventlog of the selected level: Critical, Error, Warning;
- collect of disk drives connected to device with ISCSI/SAN LUN ID;
- collect of disk volumes connected to device;
- collect of defined websited and check of expiration of SSL certificate;
- write all collected information to HTML report in filename format <HOSTNAME>-<DATE>-<TIME>.html;
- send of HTML report by email; you can use CC; all email adress are validated;
- switch (true/false) for check of expiration of SSL certificate;
- support for local and remote run of script;
- check remote device before run of main part of script;
- report still saved to same path as script;
- support for logging to file with format get-systemreport-<DATE>.txt;
- script runs only locally;
- support script create-job.ps1 for creation of job in windows task scheduler;

# EXAMPLE REPORT
- example report for Windows 10 (virtual platform / hyper-v) /examples/[Example-Report-01.html](https://github.com/michalzobec/PowerShell/blob/master/Get-SystemReport/examples/example-report-01.html);
- example report for Windows Server 2016 (virtual platform / hyper-v) /examples/[Example-Report-02.html](https://github.com/michalzobec/PowerShell/blob/master/Get-SystemReport/examples/example-report-02.html);

# QUICK START
- check and change settings in custom variables;
- test run with or without sending of mail;
- "install" as job in Windows Task Scheduler via create-job.ps1 support script;

# KNOWN PROBLEMS
- (none);

# SYSTEM REQUIREMENTS
- Windows Pro, or
- Windows Enterprise, or
- Windows Server,
- with Windows PowerShell v2+

## Developed and tested on 
- Clients: Windows 7, Windows 10,
- Servers: Windows Server 2008 R2, Windows Server 2016,
- Languages: English, Czech, Russian.

# OUTPUTS
HTML table with selected level of events in filename format <HOSTNAME>-<DATE>-<TIME>.html.

# EXAMPLE
> C:\> get-systemreport.ps1

# THANKS
- Detection of AntiVirus #1 https://www.404techsupport.com/2015/04/27/powershell-script-detect-antivirus-product-and-status/
- Detection of AntiVirus #2 http://neophob.com/2010/03/wmi-query-windows-securitycenter2/

changelog.md
# Get System Report Release History
(c) 2016-2017 Michal Zobec, ZOBEC Consulting. All Rights Reserved.  
web: www.michalzobec.cz, mail: michal@zobec.net  
License: Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)  
https://creativecommons.org/licenses/by-sa/4.0/

# DOCUMENTATION
Documentation you can find in file [ReadMe.md](https://github.com/michalzobec/PowerShell/blob/master/Get-SystemReport/readme.md).

# Release History

## version 17.07.27.1
- init version;

## version 17.07.27.2
- todo list created;

## version 17.07.27.3
- columns are filtered;
- report file name writed in format <HOSTNAME>-<DATE>-<TIME>.html;
- feature list in script description;

## version 17.07.27.4
- better code logic - included information to event block if output data does not exist (try & catch);
- better code logic - included information about code exception or another bug (try & catch);
- extended event level from Critical, Error to Critical, Error, Warning;
- added configuration value $sendmail for turn on for sending of email;
- solved problem with email sending;

## version 17.07.27.5
- added volume list with full disk size and free space size;

## version 17.07.28.1
- added support for mail CC (carbon copy);
- added validation for entered email addresses in custom values;

## version 17.08.01.1
- added eventlog id (eventid) to windows eventlog table;
- added option for turn on/off application/system events with levels critical/error/warning;
- added custom ignore list for Windows EventLog Event id;
- added hidding of hidden/service partitions in list of volumes;
- code refactoring, variables optimization and minor changes;

## version 17.08.07.1;
- renamed script to "Get System Report";
- added support for check of expiration of SSL certificate;

## version 17.08.12.1;
- added support for list of disks with ISCSI/SAN LUN ID;
- added sorting by name in tables for Disks and Volumes;

## version 17.08.13.1;
- fixed minor bugs;
- minor improvments in get-systemreport.cmd script;

## version 17.08.15.1;
- added SCSIPort, InterfaceType column for list of disks;

## version 17.08.15.2;
- added switch (true/false) for check of expiration of SSL certificate;
- new default settings of script - warning level off, check of ssl off;

## version 17.08.17.1;
- hardcoded filtering of unwanted events and partitions;
- support for local and remote run of script;

## version 17.08.21.1;
- added "installation" script for simple creation of job in windows task scheduler - create-job.cmd;
- report still saved to same path as script;
- added information on screen about path of the report;
- disk partitions filtering removed;

## version 17.08.28.1;
- email subject changed from "Daily Eventlog Report $ShortDate $HostName" to "System Report $HostName $ShortDate";
- added support for logging to file with format get-systemreport-<DATE>.txt;
- fixed problems with custom ignore list for Windows EventLog Event id

## version 17.08.28.2;
- code optimalisation for PowerShell v2;

## version 17.09.08.1;
- script now support local run only, remote run support was removed;
- more detailed information in log file;
- more detailed information in text output for manual run;
- integrated basic system report: Hardware Information, Windows Information, Network Information;
- Windows product key (in Windows Information Section) is now possibly not include to report;
- Windows Information: uptime, last boot time;
- custom variables are now in two section for custom settings and email settings;
- code refactoring, reorganization of parts of code, minor changes;

## version 17.09.08.2;
- fixed problem with loading of external libraries;
- extended logging in log file;
- extended DEBUG level logging in log file;
- extended debug information in report file;

## version 17.09.09.1;
- added Windows Computer Description for definition of device role in Report;
- extended DEBUG level logging in log file;
- extended debug information in report file;

## version 17.09.17.1;
- added time zone to Windows Information section;
- reduced DEBUG Info in Report;

## version 17.09.17.2;
- code refactoring, minor changes;
- added new section Quick Start;

## version 17.09.17.3;
- code refactoring, renamed function with unapproved verb from Validate-EmailAdress to Test-EmailAdress;

## version 17.09.17.4;
- code refactoring, added StrictMode;

## version 17.09.18.1
- improved output information;
- improved logging;
- code refactoring, detection of required script files;
- code refactoring, cleaned some parts of code;
- code refactoring, minor changes;

## version 17.09.19.1
- size will shown in GB with two decimal digits;
- physical and logical disks now filtered for "Fixed hard disk media" and "External hard disk media";
- added column free disk space in percents to Volume Information section;
- list stopped services with automatic run after start of Windows;

## version 17.10.02.1
- changed of file name format of the logfile to get-systemreport-log-<DATE>.txt;
- external libraries (ConvertTo-ProductKey, Check-SSL) was moved to subdirectory .\lib\;
- all custom variables was moved to external configuration file to subdirectory .\config\;
- HTML report will be now saved to subdirectory .\logs\;
- log file will be now saved to subdirectory .\logs\;
- example configuration file get-systemreport-config-example.ps1;

## version 17.10.03.1
- renamed configuration value SkipCheckSSL to CheckSSL;

## version 17.10.17.1
- removed CheckSSL feature;

## version 17.10.18.1
- code refactoring, better part of code for sending of email, error message to log file;
- added DEBUG information about used account to log file;

## version 17.11.05.1 (Public version on GitHub)
- code refactoring, check if exist configuration variables;
- code refactoring, checked PowerShell version for check of stopped services due incopatibility with PS 2.0;
- code refactoring, cleaned some parts of code;

create-job-old.ps1
#Requires -Version 2
<#

.SYNOPSIS
Create Job
Create job in Windows Task Scheduler for script Get System Report.
Script supports run locally.

.DESCRIPTION
Create Job
Create job in Windows Task Scheduler for script Get System Report.

FEATURES
* create new job;

KNOWNPROBLEMS
* (none);

SYSTEMREQUIREMENTS
* Windows Pro, or
* Windows Enterprise, or
* Windows Server
* Windows PowerShell v3+

Developed and tested on 
* Windows Server 2012,
* Windows 10, Windows Server 2016.

.OUTPUTS
.

.EXAMPLE
C:\> get-systemreport.ps1

.LINK
http://www.michalzobec.cz/

.NOTES
Create Job
(c) 2016-2017 Michal Zobec, ZOBEC Consulting. All Rights Reserved.
web: www.michalzobec.cz, mail: obchod@zobec.net
License: Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0) https://creativecommons.org/licenses/by-sa/4.0/

.HISTORY
version 17.08.20.1;
	* init version;
version 17.08.28.1;
	* verified on PSv2, minimal requirements is PSv3;
version 17.10.17.1;
	* changed used account to SYSTEM;

.TODO
	* (none);
	
#>

$DateTime = Get-Date -Format "dd.MM.yyyy HH:mm:ss"
$Revision = Get-Date -Format "yy.MM.dd-HH.mm"
$Scriptfile = (split-path $myinvocation.mycommand.path) + "\get-systemreport.ps1"
$TaskName = "Get System Report (inst $Revision)"
if ($Task -ne $null)
{
    Unregister-ScheduledTask -TaskName $TaskName -Confirm:$False
}
# write-host "schtasks /Create /SC daily /TN $TaskName /ST 6:00 /TR "powershell.exe -ExecutionPolicy Bypass -File `"$Scriptfile`"" /RU $env:USERDOMAIN\$env:USERNAME"
schtasks /Create /SC daily /TN "$TaskName" /ST 06:00 /TR "powershell.exe -ExecutionPolicy Bypass -File `"$Scriptfile`"" /RU $env:USERDOMAIN\$env:USERNAME


config/get-systemreport-config-example.ps1
######
# custom variables - begin
# Include Windows product key in report
$WindowsPkInReport = "True"
# ApplicationCritical
$ApplicationCritical = "True"
# ApplicationError
$ApplicationError = "True"
# ApplicationWarning
$ApplicationWarning = "True"
# SystemCritical
$SystemCritical = "True"
# ApplicationCritical
$SystemError = "True"
# SystemWarning
$SystemWarning = "True"
# hours for collect of eventlogs
$Hours = "24"
# critical free disk space in GB
$CriticalFreeDiskSpace = "22"
# skip events with event id
# Event ID 400, Message text "Exchange Server, cross-forest Availability service that can fill request for mailbox";
# Event ID 4999, Message text "Exchange Server, Watson report about to be sent to dw20.exe";
$SkipEventIDlist = (400, 4999)
# skip services name
$SkipServicesName = ("Garmin Device Interaction Service", "DoSvc")
# skip windows tasks/jobs
$SkipTaskName = ("Notifications", "SilentCleanup")
# custom variables - end
######

######
# mail settings - begin
# send email?
$SendMail = "False"
# sender email
$EmailSender = "sender@domain.ext"
# recipient email
$EmailRecipient = "recipient@domain.ext"
# carbon copy (CC) email
$EmailRecipientCC = "recipient.copy@domain.ext"
# relay mail server
$EmailRelay = "mail.relay.domain.ext"
# mail settings - end
######


create-job.ps1
#Requires -Version 3
#Requires -RunAsAdministrator
<#

.SYNOPSIS
Create Job
Create job in Windows Task Scheduler for script Get System Report.
Script supports run locally.

.DESCRIPTION
Create Job
Create job in Windows Task Scheduler for script Get System Report.

FEATURES
* create new job;

KNOWNPROBLEMS
* (none);

SYSTEMREQUIREMENTS
* Windows Pro, or
* Windows Enterprise, or
* Windows Server
* Windows PowerShell v3+

Developed and tested on 
* Windows 7, Windows Server 2008 R2,
* Windows 10, Windows Server 2016.

.OUTPUTS
.

.EXAMPLE
C:\> get-systemreport.ps1

.LINK
http://www.michalzobec.cz/

.NOTES
Create Job
(c) 2016-2017 Michal Zobec, ZOBEC Consulting. All Rights Reserved.
web: www.michalzobec.cz, mail: obchod@zobec.net
License: Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0) https://creativecommons.org/licenses/by-sa/4.0/

.HISTORY
version 17.08.20.1;
	* init version;
version 17.08.28.1;
	* verified on PSv2, minimal requirements is PSv3;
version 17.09.08.1;
	* added trigger with run of script after Windows start;
	* add support for older systems Windows 7, Windows Server 2008 R2;
version 17.09.18.1;
	* added Requires RunAsAdministrator statement;
version 17.10.18.1;
	* changed settings for used account;

.TODO
	* (none);

#>

$DateTime = Get-Date -Format "dd.MM.yyyy HH:mm:ss"
$Revision = Get-Date -Format "yy.MM.dd-HH.mm"
$Scriptfile = (split-path $myinvocation.mycommand.path) + "\get-systemreport.ps1"
$Action = New-ScheduledTaskAction -Execute "powershell.exe" -Argument "-ExecutionPolicy Bypass -File `"$Scriptfile`""
$Triggers = @()
$Triggers += New-ScheduledTaskTrigger -Daily -At 06:00
$Triggers += New-ScheduledTaskTrigger -AtStartup
# $Principal = New-ScheduledTaskPrincipal -UserId SYSTEM -LogonType ServiceAccount -RunLevel Highest
# $Principal = New-ScheduledTaskPrincipal -RunLevel Highest -LogonType S4U -UserId $env:USERDOMAIN\$env:USERNAME
$Principal = New-ScheduledTaskPrincipal -LogonType S4U -UserId $env:USERDOMAIN\$env:USERNAME
$TaskName = "Get System Report (inst $Revision)"
$Description = "Get System Report PS daily script`ninstalled $DateTime"
$Task = Get-ScheduledTask -TaskName $TaskName -ErrorAction SilentlyContinue
if ($Task -ne $null)
{
    Unregister-ScheduledTask -TaskName $TaskName -Confirm:$False
}
Register-ScheduledTask -Action $Action -Principal $Principal -Trigger $Triggers -TaskName $TaskName -Description $Description